Use the dataset `Normalised_merged_GE_25scaffolds_ALLTISSUES.txt` (produced in section **04.GE_quantification_normalisation.md**) contained in the online Supplementary Material **"Dataset S4: Gene expression and dosage compensation/datasets for tissue-specific expression and Spearman correlation"** for this analysis. 

In R
```ruby
### 1. Load packages

library(ggplot2)
library(ggpubr)
library(rcompanion)
library(stringr)


### 2. Load dataset and keep only useful columns

GE <- read.table("~/PATH/Normalised_merged_GE_25scaffolds_ALLTISSUES.txt", head=T, sep="")
head(GE)
str(GE)
GE$chromosome <- as.factor(GE$chromosome)
summary(GE$chromosome)
GE_AVG_df = GE[, c(1,3, 22:25)]
head(GE_AVG_df)


### 3. Tissue-specific expression

## define TPM cuttoffs under which a gene is considered as not-expressed
cuttoff_expressed <- 1
cuttoff_notexpressed <- 0.5

## create a new column for each tissue with level of specificity ("1"= specicific, "0" = not specifici)

#### TESTES
for(i in 1:nrow(GE_AVG_df)){  
  if(GE_AVG_df$AVG_TESTES[i] > cuttoff_expressed
     & GE_AVG_df$AVG_HEADS[i] < cuttoff_notexpressed
     & GE_AVG_df$AVG_OVARIES[i] < cuttoff_notexpressed
     & GE_AVG_df$AVG_CARCASSES[i] < cuttoff_notexpressed){   
    GE_AVG_df$TESTES_SPE[i] <- 1  
  } else {
    GE_AVG_df$TESTES_SPE[i] <- 0
  }
}

# number of genes with testes-specific expression:
sum(GE_AVG_df$TESTES_SPE)

#### OVARIES
for(i in 1:nrow(GE_AVG_df)){  
  if(GE_AVG_df$AVG_OVARIES[i] > cuttoff_expressed
     & GE_AVG_df$AVG_HEADS[i] < cuttoff_notexpressed
     & GE_AVG_df$AVG_TESTES[i] < cuttoff_notexpressed
     & GE_AVG_df$AVG_CARCASSES[i] < cuttoff_notexpressed){   
    GE_AVG_df$OVARIES_SPE[i] <- 1  
  } else {
    GE_AVG_df$OVARIES_SPE[i] <- 0
  }
}

# number of genes with ovaries-specific expression:
sum(GE_AVG_df$OVARIES_SPE)



#### HEADS
for(i in 1:nrow(GE_AVG_df)){  
  if(GE_AVG_df$AVG_HEADS[i] > cuttoff_expressed
     & GE_AVG_df$AVG_OVARIES[i] < cuttoff_notexpressed
     & GE_AVG_df$AVG_TESTES[i] < cuttoff_notexpressed
     & GE_AVG_df$AVG_CARCASSES[i] < cuttoff_notexpressed){   
    GE_AVG_df$HEADS_SPE[i] <- 1  
  } else {
    GE_AVG_df$HEADS_SPE[i] <- 0
  }
}

# number of genes with head-specific expression:
sum(GE_AVG_df$HEADS_SPE)


#### CARCASSES
for(i in 1:nrow(GE_AVG_df)){  
  if(GE_AVG_df$AVG_CARCASSES[i] > cuttoff_expressed
     & GE_AVG_df$AVG_OVARIES[i] < cuttoff_notexpressed
     & GE_AVG_df$AVG_TESTES[i] < cuttoff_notexpressed
     & GE_AVG_df$AVG_HEADS[i] < cuttoff_notexpressed){   
    GE_AVG_df$CARCASSES_SPE[i] <- 1  
  } else {
    GE_AVG_df$CARCASSES_SPE[i] <- 0
  }
}

# number of genes with carcass-specific expression:
sum(GE_AVG_df$CARCASSES_SPE)



### 4. Autosomal vs X-linkage assignement, per tissue

## create subsets for the 4 tissues containing chromosomal location and tissue-specificity
testes_GE_AVG_df = GE_AVG_df[ ,c(2,7)]
ovaries_GE_AVG_df = GE_AVG_df[ ,c(2,8)]
heads_GE_AVG_df = GE_AVG_df[ ,c(2,9)]
carcasses_GE_AVG_df = GE_AVG_df[ ,c(2,10)]
# sanity checks
head(testes_GE_AVG_df)
head(ovaries_GE_AVG_df)
head(heads_GE_AVG_df)
head(carcasses_GE_AVG_df)


## Assign X and autosomal linkage to each of the 4 subsets

#### TESTES
for(i in 1:nrow(testes_GE_AVG_df)){  
  if(testes_GE_AVG_df$chromosome[i] == "A"){   
    testes_GE_AVG_df$chr_tissue[i] <- "A_testes"  
  } else {
    testes_GE_AVG_df$chr_tissue[i] <- "X_testes"
  }
}

# sanity checks
str(testes_GE_AVG_df$chr_tissue)
testes_GE_AVG_df$chr_tissue <- as.factor(testes_GE_AVG_df$chr_tissue)
summary(testes_GE_AVG_df$chr_tissue)


#### OVARIES
for(i in 1:nrow(ovaries_GE_AVG_df)){  
  if(ovaries_GE_AVG_df$chromosome[i] == "A"){   
    ovaries_GE_AVG_df$chr_tissue[i] <- "A_ovaries"  
  } else {
    ovaries_GE_AVG_df$chr_tissue[i] <- "X_ovaries"
  }
}

# sanity checks
str(ovaries_GE_AVG_df$chr_tissue)
ovaries_GE_AVG_df$chr_tissue <- as.factor(ovaries_GE_AVG_df$chr_tissue)
summary(ovaries_GE_AVG_df$chr_tissue)


#### HEADS
for(i in 1:nrow(heads_GE_AVG_df)){  
  if(heads_GE_AVG_df$chromosome[i] == "A"){   
    heads_GE_AVG_df$chr_tissue[i] <- "A_heads"  
  } else {
    heads_GE_AVG_df$chr_tissue[i] <- "X_heads"
  }
}

# sanity checks
str(heads_GE_AVG_df$chr_tissue)
heads_GE_AVG_df$chr_tissue <- as.factor(heads_GE_AVG_df$chr_tissue)
summary(heads_GE_AVG_df$chr_tissue)


#### CARCASSES
for(i in 1:nrow(carcasses_GE_AVG_df)){  
  if(carcasses_GE_AVG_df$chromosome[i] == "A"){   
    carcasses_GE_AVG_df$chr_tissue[i] <- "A_carcasses"  
  } else {
    carcasses_GE_AVG_df$chr_tissue[i] <- "X_carcasses"
  }
}

# sanity checks
str(carcasses_GE_AVG_df$chr_tissue)
carcasses_GE_AVG_df$chr_tissue <- as.factor(carcasses_GE_AVG_df$chr_tissue)
summary(carcasses_GE_AVG_df$chr_tissue)
head(carcasses_GE_AVG_df)



## Rename the columns
colnames(heads_GE_AVG_df)<-c("chromosome","specificity", "chr_tissue") 
colnames(testes_GE_AVG_df)<-c("chromosome","specificity", "chr_tissue") 
colnames(ovaries_GE_AVG_df)<-c("chromosome","specificity", "chr_tissue") 
colnames(carcasses_GE_AVG_df)<-c("chromosome","specificity", "chr_tissue") 

# Merge the 4 datasets 
spe_df <- do.call("rbind", list(ovaries_GE_AVG_df, testes_GE_AVG_df, heads_GE_AVG_df, carcasses_GE_AVG_df))
str(spe_df)
head(spe_df)

# Organise the levels of chr_tissue variable for the plot
spe_df$chr_tissue <- factor(spe_df$chr_tissue, levels = c("A_ovaries",
                                                          "X_ovaries", 
                                                          "A_testes",
                                                          "X_testes", 
                                                          "A_heads",
                                                          "X_heads", 
                                                          "A_carcasses",
                                                          "X_carcasses"))

### 5. Calculate tissue-specificity percentages
Specific_genes <- tapply(spe_df$specificity, spe_df$chr_tissue, FUN = function(x) sum(x))
Non_Specific_genes <- tapply(spe_df$specificity, spe_df$chr_tissue, FUN = function(x) length(x)-sum(x))
Specificity_perc <- tapply(spe_df$specificity, spe_df$chr_tissue, FUN = function(x) sum(x)/length(x)*100)
Chr_tissue <- names(Specificity_perc)
Chromosome <- substr(Chr_tissue,1,1)

# put all of this info in a new dataframe called perc_df
perc_df <- data.frame(Chromosome, Chr_tissue, Specific_genes, Non_Specific_genes, Specificity_perc)
perc_df


# do the stats to compare the number of specific vs non-specific on the X and autosomes - per tissue
df_for_chi_square = perc_df [, c(3:4)]
df_for_chi_square

### 6. perform a chi-square tests
Matriz = data.matrix(df_for_chi_square)
stat.test <- pairwiseNominalIndependence(Matriz,
                                         fisher = FALSE,
                                         gtest  = FALSE,
                                         chisq  = TRUE,
                                         method = "fdr")

stat.test

# retain only the comparisons needed for the plot: A_ovaries : X_ovaries, A_testes : X_testes, A_heads : X_heads, A_carcasses : X_carcasses
stat.test.rows = stat.test [c(1,14,23,28), ]
head(stat.test.rows)
# tranform the dataframe to make it readable by the plot : split the comparison column into 2 new variables "groups"
stat.test.rows[c('group1', 'group2')] <- str_split_fixed(stat.test.rows$Comparison, ' : ', 2)
# delete the comparison column
stat.test.rows = stat.test.rows[, -1]
# re-organise the dataframe
stat.test.rows2 = stat.test.rows[c("group1", "group2", "p.Chisq" , "p.adj.Chisq")]

# make groups as factors
stat.test.rows2$group1 <- as.factor(stat.test.rows2$group1)
stat.test.rows2$group2 <- as.factor(stat.test.rows2$group2)

# out of the 2 p-value columns, keep adjusted
stat.test.rows2 = stat.test.rows2[, -3]

# rename columms
colnames(stat.test.rows2)<-c("group1", "group2", "p") 
head(stat.test.rows2)

# adjust factor levels for the plots
perc_df$Chr_tissue <- factor(perc_df$Chr_tissue, levels =  c("A_ovaries",
                                                             "X_ovaries",
                                                             "A_testes",
                                                             "X_testes",
                                                             "A_heads",
                                                             "X_heads",
                                                             "A_carcasses",
                                                             "X_carcasses"))


### 6. Plot
PLOT <- ggplot(data = perc_df, aes(y=Specificity_perc, x=Chr_tissue, fill=Chromosome)) + 
  
  geom_bar(data = perc_df, position="dodge", stat="identity") +
  
  scale_fill_manual(values = c("#969696","#ffd92f"), labels = c('Autosomes', 'X')) +
  
  coord_cartesian(ylim = c(0, 8))+
  ylab("Percentage of genes") +
  scale_y_continuous(expand = c(0, 0), limits = c(0, NA))+
  
  xlab("Tissues") + 
  
  ggtitle("(e) Tissue-specific expression") +
  
  stat_pvalue_manual(data=stat.test.rows2, 
                     label.size=3,
                     inherit.aes=FALSE, 
                     y.position=c(5.6, 7.7, 2.8, 2.7),
                     label = "p = {p}", 
                     fontface = "bold", 
                     bracket.size = 0.5, 
                     tip.length = 0)+
  
  theme(
    legend.position = c(.99, .99),
    legend.justification = c("right", "top"),
    legend.text=element_text(size=8),
    legend.title=element_blank(),
    
    axis.line = element_line(colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black",fill = NA,size = 1),
    panel.background = element_blank(),
    
    plot.title = element_text(hjust = 0, size=14, face = "bold"),
    
    axis.text.y = element_text(size=12),
    axis.title.y = element_text(margin=margin(r=7), size = 13, face ="bold"),
    
    axis.text.x = element_text(vjust = 0.2, size = 12, face ="bold", color = "black"),
    axis.title.x = element_blank())+
  
  scale_x_discrete(limits = c("A_ovaries", "X_ovaries","A_testes", "X_testes","A_heads", "X_heads", "A_carcasses", "X_carcasses"), 
                  breaks = c("A_ovaries","X_ovaries","A_testes","X_testes","A_heads","X_heads", "A_carcasses", "X_carcasses"),
                  labels=c('ovaries','ovaries', 'testes','testes', 'heads','heads','carcasses','carcasses'))


PLOT

#To save the plot in high quality on computer:
#jpeg("tissue_specific_plot.jpg", width = 4, height = 5, units = "in", res = 400)
#PLOT
#dev.off()
```
# detach packages
detach("package:ggpubr", unload=TRUE)
detach("package:ggplot2", unload=TRUE)
detach("package:rcompanion", unload=TRUE)
detach("package:stringr", unload=TRUE)
``` 
