Files to use and source directories in Seafile:

### For Heads
Use dataset **Normalised_HEADS_GE_merged_25_scaffolds_filter_0.5.txt** from **"Dataset S4: Gene expression and dosage compensation/dosage_compensation_datasets/HEADS"** for this analysis. 
### For Gonads
Use dataset **Normalised_GONADS_GE_merged_25_scaffolds_filter_0.5.txt** from **"Dataset S4: Gene expression and dosage compensation/dosage_compensation_datasets/GONADS"** for this analysis. 
### For Carcasses
Use dataset **Normalised_CARCASSES_GE_merged_25_scaffolds_filter_0.5.txt** from **"Dataset S4: Gene expression and dosage compensation/dosage_compensation_datasets/CARCASSES"** for this analysis. 

Go to "PanorpaX/Gene_expression_analysis/04.GE_quantification_normalisation.md" to see how these datasets were generated.

In R (eg for heads):
```ruby
# 1. load packages
library(ggplot2)
library(ggsignif)
library(ggpubr)


# 2. Load dataset (normalised expression summary file filtered for GE > 0.5 TPM)
GE<- read.table("~/Documents/MECOPTERA/post_hack_check_GE/25scaffolds_032023/GE_HEADS/Normalised_HEADS_GE_merged_25_scaffolds_filter_0.5.txt", head=T, sep="")
head(GE)
str(GE) 

# 3. create a new column containing gene location categories that will be more helpful for the plot
for(i in 1:nrow(GE)){  
  if(GE$chromosome[i] == "A"){   
    GE$chromo[i] <- "Autosomal genes"  
  } else {
    GE$chromo[i] <- "X-linked genes"
  }
}

## sanity check
head(GE)
str(GE) 

# 4. create 2 subset: 1 male and 1 female
df_male_GE = GE[ ,c(3,10,12)]
df_female_GE = GE[ ,c(3,11,12)]
head(df_male_GE)
head(df_female_GE)

## 4.1. create a new column in each subset containing sex & gene location categories that will be more helpful for the plot

# males
for(i in 1:nrow(df_male_GE)){  
  if(df_male_GE$chromosome[i] == "A"){   
    df_male_GE$sex_chr[i] <- "Male_auto"  
  } else {
    df_male_GE$sex_chr[i] <- "Male_X"
  }
}

#females
for(i in 1:nrow(df_female_GE)){  
  if(df_female_GE$chromosome[i] == "A"){   
    df_female_GE$sex_chr[i] <- "Fem_auto"  
  } else {
    df_female_GE$sex_chr[i] <- "Fem_X"
  }
}

#sanity checks
summary(df_male_GE$sex_chr)
str(df_male_GE)
summary(df_female_GE$sex_chr)
str(df_female_GE)


# 4.2. delete the "chromosome" column in each subset
df_male_GE = df_male_GE[ ,-1]
df_female_GE = df_female_GE[ ,-1]
str(df_male_GE)
str(df_female_GE)


# 5. Merge the 2 subsets together

## 5.1. first rename the "MALE_AVG" and "FEM_AVG" gene expression columns in each dataset by the same thing "AVG_GE"
colnames(df_female_GE)<-c("AVG_GE", "chromosome", "sex_chr") 
colnames(df_male_GE)<-c("AVG_GE", "chromosome", "sex_chr") 
head(df_female_GE)
head(df_male_GE)

## 5.2. Merge
avg_GE_df <- do.call("rbind", list(df_male_GE, df_female_GE))
str(avg_GE_df)

# 6. create a new column called sex in the merged dataset
for(i in 1:nrow(avg_GE_df)){  
  if(avg_GE_df$sex_chr[i] == "Fem_auto"){   
    avg_GE_df$sex[i] <- "Female" 
  } else if(avg_GE_df$sex_chr[i] == "Fem_X"){   
    avg_GE_df$sex[i] <- "Female"
  } else {
    avg_GE_df$sex[i] <- "Male"
  }
}


# 5.3. check that everything is alright with the new variables created
avg_GE_df$sex <- as.factor(avg_GE_df$sex)
avg_GE_df$sex_chr <- as.factor(avg_GE_df$sex_chr)
avg_GE_df$chromosome <- as.factor(avg_GE_df$chromosome)
summary(avg_GE_df$sex)
str(avg_GE_df)
summary(avg_GE_df)

avg_GE_df$sex_chr <- factor(avg_GE_df$sex_chr, levels = c("Fem_auto","Male_auto", "Fem_X", "Male_X"))
which(is.na(avg_GE_df$sex_chr)) # check that no NAs are produced

# How many autosomal and X-linked genes do we have?
summary(avg_GE_df$sex_chr)
unique(avg_GE_df$sex_chr)

# 6. calculate log2(avg_GE)
avg_GE_df$LOG_AVG_GE <- log2(avg_GE_df$AVG_GE)

# 7. calculate the median of LOG_AVG_GE for all 4 sex_chr categories
MALE_Auto_median <- median(avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Male_auto")])
MALE_Auto_median 
FEM_Auto_median <- median(avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Fem_auto")])
FEM_Auto_median 
MALE_X_median <- median(avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Male_X")])
MALE_X_median 
FEM_X_median <- median(avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Fem_X")])
FEM_X_median 


# 8.STATS - WILCOXON TESTS 
### Fem_auto to Male_auto
wilcox.test(avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Fem_auto")], avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Male_auto")])

### Fem_X to Male_X
wilcox.test(avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Fem_X")], avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Male_X")])

### Fem_auto to Fem_X
wilcox.test(avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Fem_auto")], avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Fem_X")])

### Male_auto to Male_X
wilcox.test(avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Male_auto")], avg_GE_df$LOG_AVG_GE[which(avg_GE_df$sex_chr=="Male_X")])



# 9. PLOTS 

# 9.1. Dosage compensation plot 
# know the margins of the plot to help define the y-axis limites
max(avg_GE_df$LOG_AVG_GE)
min(avg_GE_df$LOG_AVG_GE)

PLOT1 <- ggplot(avg_GE_df, aes(x = sex_chr, y = LOG_AVG_GE, fill = sex)) +
  
  scale_fill_manual(values = c("#00718B","#7CCBA2"), labels = c('Female', 'Male')) +
  
  coord_cartesian(ylim = c(-1, 12)) +
  ylab("Log2 (TPM)") +
  
  ggtitle("(b) Heads") +
  
  geom_signif(comparisons = list(c("Fem_auto","Male_auto"), c("Fem_X","Male_X"), c("Fem_auto","Fem_X"), c("Male_auto","Male_X")),
              map_signif_level = TRUE,
              y_position = c(9.8, 9.8, 10.5, 11.2),, 
              vjust = .1,
              tip_length = 0.005, 
              size = 0.5, 
              textsize = 4) + 
  
  geom_boxplot(outlier.shape = NA, notch = TRUE) + 
  
  geom_hline(yintercept = FEM_Auto_median, linetype="dashed", color = "#525252", size = 0.7)+
  
  theme(
    legend.position = "bottom",
    legend.text = element_text(size=12),
    legend.title = element_blank(),
    legend.box.background = element_rect(color="black", size=0.5),
    
    axis.line = element_line(colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black",fill = NA,size = 1),
    panel.background = element_blank(),
    
    plot.title = element_text(hjust = 0, size=14, face = "bold"),
    
    axis.text.y = element_text(size=12),
    axis.title.y = element_text(margin=margin(r=3), size = 13, face ="bold"),
    
    axis.text.x = element_text(vjust = 0.2, size = 12, face ="bold", color = "black"),
    
    axis.title.x = element_blank())

PLOT1




# 9.2. Male/Female ratio 

## generate a ratio column in the dataframe
GE$MF_ratio_log <- log2( GE$MALE_AVG / GE$FEM_AVG )

## STATS - WILCOXON TESTS 
wilcox.test(GE$MF_ratio_log[which( GE$chromosome == "A")], GE$MF_ratio_log[which(GE$chromosome =="X")])


### 9.2.1. Plot

PLOT2 <- ggplot(GE, aes(x = chromosome, y = MF_ratio_log, fill = chromosome)) +
  
  scale_fill_manual(values = c( "#969696","#ffd92f")) +
  
  geom_signif(comparisons = list( c("A","X")), 
              map_signif_level = TRUE,
              y_position = 0.27, 
              vjust = .1,
              tip_length = 0.001, 
              size = 0.5, 
              textsize = 4) +
  
  geom_boxplot(outlier.shape = NA, notch = TRUE) + 
  
  ggtitle("(d) Heads")+
  
  coord_cartesian(ylim = c(-1, 1.2))  +
  
  ylab("Log2(M/F)") +
  
  geom_hline(yintercept=0, linetype="dashed", color = "black", size=0.7)+
  
  theme(
    legend.position = "bottom",
    legend.text = element_text(size=12),
    legend.title = element_blank(),
    legend.box.background = element_rect(color="black", size=0.5),
    
    axis.line = element_line(colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black",fill = NA,size = 1),
    panel.background = element_blank(),
    
    plot.title = element_text(hjust = 0, size=14, face = "bold"),
    
    axis.text.y = element_text(size=12),
    axis.title.y = element_text(margin=margin(r=3), size = 13, face ="bold"),
    
    axis.text.x = element_text(vjust = 0.2, size = 12, face ="bold", color = "black"),
    axis.title.x = element_blank())+

scale_x_discrete(limits = c("A", "X"), 
                 breaks = c("A", "X"))

PLOT2


### arrange plots in the same window ##
plots <- ggarrange(PLOT1, PLOT2, nrow=1, ncol=2, widths=c(1.8, 1))
plots

## save plots
#jpeg("heads_DC_plots_25_scaffolds.jpg", width = 6, height = 5, units = "in", res = 400)
#plots
#dev.off()

# detach packages
detach("package:ggpubr", unload=TRUE)
detach("package:ggplot2", unload=TRUE)
detach("package:ggsignif", unload=TRUE)
```
Repeat for the 2 other tissues: gonads and carcasses
