This analysis requires 3 datasets previously produced:

* (1) the file indicating the genomic location (autosomes vs. X) of each transcripts: `trans_location25.txt`. This file canbe found in the online Supplementary Material in "PanorpaX/Dataset S2: Transcriptome assembly and transcripts genomic location/".

* (2) the tissue-specific list of differentially expressed genes between males and females `DEgenes_GONAD_samples.txt`. This file was produced by Sleuth (see section **04.GE_quantification_normalisation.md** for pipeline). This file can be found in the online Supplementary Material in "PanorpaX/Dataset S4: Gene expression and dosage compensation/sex_biased_expression/"

* (3) the gene expression summary files for this tissue `NonNormalised_merged_GE_25scaffolds_GONADS.txt`, also produced by Sleuth (see section **04.GE_quantification_normalisation.md**). This dataset can be found in the online Supplementary Material in "PanorpaX/Dataset S4: Gene expression and dosage compensation/GONADS/".

*The same analysis can be repeated for heads and carcasses using the HEADS- and CARCASSES-specific files for (2) and (3).*

# 1. Filter for differentially expressed genes with q-values < 0.05

Example with gonads:

```ruby
## LOAD chrom loc dataset
chrom_location <- read.table("~/PATH/trans_location25.txt", head=T, sep=",") # 12357 transcripts: 1274 X, and 11083 autosomal
head(chrom_location)
colnames(chrom_location)<-c("Transcript", "Scaffold", "chromosome") 
str(chrom_location$chromosome)
chrom_location$chromosome <- factor(chrom_location$chromosome, levels = c("A", "X"))
levels(chrom_location$chromosome)
summary(chrom_location$chromosome)

### LOAD diff exp genes
df <- read.table("~/PATH/DEgenes_GONAD_samples.txt", head=T, sep="")
head(df)
str(df)

# merge the 2 datasets
merged_df <-merge(chrom_location,df, by.x="Transcript", by.y="target_id") 
head(merged_df)
summary(merged_df$chromosome)


# subset transcripts with qval < 0.05
library(plyr)
library(dplyr)

merged_df_qsignif <- merged_df %>% filter(qval < 0.05)
summary(merged_df_qsignif$chromosome)

# keep only interesting columns
merged_df_qsignif = merged_df_qsignif[, c(1,3)]
head(merged_df_qsignif)

# write a new file
write.table(merged_df_qsignif ,"~/PATH/GONADS_signif_GEgenes_merged_chrom_loc.txt")
```

# 2. Normalise expression summary

In R:

```ruby
# open gene expression dataset to be normalised 
expf <- read.table("NonNormalised_merged_GE_25scaffolds_GONADS.txt", head=T, sep="")
head(expf)
# delete the column(s) that are not GE data (eg. gene name or chromosome location)
expf_1 <- expf[, -c(1:3)]
head(expf_1)
# create a new matrix that has a number of columns and rows that match the expression gene dataset
bolFMat_1<-as.matrix(expf_1, nrow = nrow(expf_1), ncol = ncol(expf_1))

# load the data normalisation package
library(NormalyzerDE)

# run data normalisation
expf_2<-performQuantileNormalization(bolFMat_1, noLogTransform = T)

# save the matrix containing normalised data as a dataframe
expf_2<-as.data.frame(expf_2)
head(expf_2)

# reintroduce the gene name column into the dataframe
expf_2 <- data.frame(expf[, c(1:3)], expf_2)
# check that the gene names match between expf and expf_2
head(expf)
head(expf_2)

# Make GE averages per sex
expf_2$MALE_AVG <- rowMeans(expf_2[ ,c("MALE1", "MALE2", "MALE3")])
expf_2$FEM_AVG <- rowMeans(expf_2[ ,c("FEM1", "FEM2", "FEM3")])

# Make sure that all these new variables are Numeric and contain no NA (although no essential due to next step)
expf_2$MALE_AVG <- as.numeric(expf_2$MALE_AVG)
expf_2$FEM_AVG <- as.numeric(expf_2$FEM_AVG)
which(is.na(expf_2$MALE_AVG))
which(is.na(expf_2$FEM_AVG))
head(expf_2)

# delete the column(s) that are not GE data (eg. gene name or chromosome location)
expf_3 <- expf_2[, -c(1:9)]
head(expf_3)

# create a new matrix that has a number of columns and rows that match the expression gene dataset
bolFMat_3 <- as.matrix(expf_3, nrow = nrow(expf_3), ncol = ncol(expf_3))

# run data normalisation
expf_4 <- performQuantileNormalization(bolFMat_3, noLogTransform = T)

# save the matrix containing normalised data as a dataframe
expf_4 <- as.data.frame(expf_4)

# reintroduce the gene name column into the dataframe
expf_4 <- data.frame(expf_2[, c(1:9)], expf_4)

# check that the gene names match between before and after the second transformation
head(expf_3)
head(expf_4)

# save the new dataset of normalised data
write.table(expf_4, file = "Normalised_GONADS_GE_merged_25_scaffolds_NO_filter_30052023.txt", quote=F)
```

# 3. Sex-biased gene expression analysis

R code:
```ruby
# 1. load packages
library(ggplot2)
library(ggpubr)
library(rcompanion)
library(stringr)
library(dplyr)
library(EnvStats)

# 2. load datasets

# 2.1 Signif. DE genes for the tissue (i.e. Sleuth output)
df_signif <- read.table("~/PATH/GONADS_signif_GEgenes_merged_chrom_loc.txt", head=T, sep="")
head(df_signif)
df_signif$chromosome <- as.factor(df_signif$chromosome)
summary(df_signif$chromosome)
length(df_signif$chromosome)

# 2.2. Normalised GE for this tissue without TPM filter
GE <-read.table("~/PATH/Normalised_GONADS_GE_merged_25_scaffolds_NO_filter_30052023.txt", head=T, sep="")
head(GE)
str(GE)
GE$chromosome <- as.factor(GE$chromosome)
summary(GE$chromosome)
length(GE$chromosome)

# extract the number of autosomal and X.linked genes for percentage analysis further down
total_Auto_genes <- nrow(GE[GE$chromosome=="A",])
total_X_genes <- nrow(GE[GE$chromosome=="X",])

# keep only interesting columns from GE dataset
GE = GE[, c(1,10,11)]
head(GE)

# 3. merge GE and signif DEgenes
merged <-merge(GE, df_signif, by.x="Transcript", by.y="Transcript") 
head(merged)
summary(merged$chromosome)
str(merged)

# 4. apply GE > 0.5 TPM filter on both male and female averaged GE
merged <- merged %>% filter(MALE_AVG > 0.5, FEM_AVG > 0.5)

# sanity checks
which(merged$MALE_AVG < 0.5)
which(merged$FEM_AVG < 0.5)
# how many autosomal and X-linked genes are left after filtering?
summary(merged$chromosome)

####################################################################
######################### N-FOLD ANALYSIS ##########################
####################################################################

# 5. calculate nfold increses in gene expression on the 0.5 TPM filtered dataset

## if male GE > female GE 

for(i in 1:nrow(merged)){  
  if(merged$MALE_AVG[i] > merged$FEM_AVG[i] 
  ){   
    merged$Male_nfold[i] <- (merged$MALE_AVG[i]/merged$FEM_AVG[i])
  } else {
    merged$Male_nfold[i] <- 0
  }
}


## if female GE > male GE

for(i in 1:nrow(merged)){  
  if(merged$MALE_AVG[i] < merged$FEM_AVG[i] 
  ){   
    merged$Fem_nfold[i] <- (merged$FEM_AVG[i]/merged$MALE_AVG[i])
  } else {
    merged$Fem_nfold[i] <- 0
  }
}

# sanity checks
head(merged)
str(merged)


# 6. create 2 sex-specific subsets containing chromosomal location and nfold and Assign X and autosomal linkage to each subset
MALE_nfold_df = merged[ ,c(4,5)]
head(MALE_nfold_df)
FEMALE_nfold_df  = merged[ ,c(4,6)]
head(FEMALE_nfold_df)

## MALE_nfold_df
for(i in 1:nrow(MALE_nfold_df)){  
  if(MALE_nfold_df$chromosome[i] == "A"){   
    MALE_nfold_df$chr_sex[i] <- "Auto_MALES"  
  } else {
    MALE_nfold_df$chr_sex[i] <- "X_MALES"
  }
}

# sanity checks
str(MALE_nfold_df$chr_sex)
MALE_nfold_df$chr_sex <- as.factor(MALE_nfold_df$chr_sex)
summary(MALE_nfold_df$chr_sex)
head(MALE_nfold_df)

## FEMALE_nfold_df
for(i in 1:nrow(FEMALE_nfold_df)){  
  if(FEMALE_nfold_df$chromosome[i] == "A"){   
    FEMALE_nfold_df$chr_sex[i] <- "Auto_FEMALES"  
  } else {
    FEMALE_nfold_df$chr_sex[i] <- "X_FEMALES"
  }
}

# sanity checks
str(FEMALE_nfold_df$chr_sex)
FEMALE_nfold_df$chr_sex <- as.factor(FEMALE_nfold_df$chr_sex)
summary(FEMALE_nfold_df$chr_sex)
head(FEMALE_nfold_df)


# 7. To assess the distribution of nfolds, filter against genes for which nfold = 0 (i.e. unbiased)
MALE_nfold_filtered_0 <- MALE_nfold_df %>% filter(Male_nfold > 0)
FEMALE_nfold_filtered_0 <- FEMALE_nfold_df %>% filter(Fem_nfold > 0)

# Sanity checks (no nfold values of 0 in either male of female subsets)
head(MALE_nfold_filtered_0)
head(FEMALE_nfold_filtered_0)
which(MALE_nfold_filtered_0$Male_nfold < 0.001)
which(FEMALE_nfold_filtered_0$Fem_nfold < 0.001)
summary(MALE_nfold_filtered_0$chromosome)
summary(FEMALE_nfold_filtered_0$chromosome)


# 7. Rename the columns in each of the two subsets and then merge them
colnames(MALE_nfold_filtered_0)<-c("chromosome","nfold", "chr_sex") 
colnames(FEMALE_nfold_filtered_0)<-c("chromosome","nfold", "chr_sex") 
head(FEMALE_nfold_filtered_0)

nfold_df <- do.call("rbind", list(FEMALE_nfold_filtered_0, MALE_nfold_filtered_0))

# sanity checks
str(nfold_df)
head(nfold_df)
summary(nfold_df$chr_sex)


#### 8. PLOT nfold distribution ####

nfold_plot <- ggplot(nfold_df, aes(x = chr_sex, y = log2(nfold), fill = chromosome)) +
  
  scale_fill_manual(values = c("#969696","#ffd92f"), labels = c('Autosomes', 'X')) +
  
  coord_cartesian(ylim = c(-1, 10)) +
  ylab("Log2 (nfold)") +
  
  ggtitle("Gonads: Distribution of nfold values of DE genes \n(q < 0.05, TPM > 0.5) \n(test: wilcoxon)") +
  
  geom_signif(comparisons = list(c("Auto_FEMALES","X_FEMALES"), c("Auto_MALES","X_MALES"), c("Auto_FEMALES","Auto_MALES"), c("X_FEMALES","X_MALES")),
              map_signif_level = TRUE,
              y_position = c(6, 7.5, 7, 8.2),
              vjust = .1,
              tip_length = 0.005, 
              size = 0.5, 
              textsize = 4) + 
  
  geom_boxplot(outlier.shape = NA, notch = TRUE) + 
  
  stat_n_text(size = 4, y.pos = -0.5)+
  
  theme(
    legend.position = "none",
    
    axis.line = element_line(colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black",fill = NA,size = 1),
    panel.background = element_blank(),
    
    plot.title = element_text(hjust = 0, size=7, face = "bold"),
    
    axis.text.y = element_text(size=12),
    axis.title.y = element_text(margin=margin(r=3), size = 13, face ="bold"),
    
    axis.text.x = element_text(vjust = 1, hjust = 1, size = 10, face ="bold", color = "black", angle = 45),
    
    axis.title.x = element_blank())+
  
  scale_x_discrete(limits = c("Auto_FEMALES","X_FEMALES","Auto_MALES","X_MALES"),
                   breaks = c("Auto_FEMALES","X_FEMALES","Auto_MALES","X_MALES"),
                   labels=c('female_biased','female_biased', 'male_biased','male_biased'))


nfold_plot




#######################################################################
######################## SEX-BIAS ANALYSIS ############################
#######################################################################

# Used the "merged" dataset

# 5. set a nfold cutoff
nfold_cuttoff = 2

# 6. create 2 new variables to calculate the percentage of male- and female-biased genes

## male-biased_genes
for(i in 1:nrow(merged)){  
  if(merged$Male_nfold[i] > nfold_cuttoff
  ){   
    merged$Male_biased[i] <- 1
  } else {
    merged$Male_biased[i] <- 0
  }
}

## female-biased genes
for(i in 1:nrow(merged)){  
  if(merged$Fem_nfold[i] > nfold_cuttoff
  ){   
    merged$Female_biased[i] <- 1
  } else {
    merged$Female_biased[i] <- 0
  }
}

# sanity check
head(merged)


# 7. create sex-specific subsets containing chromosomal location and sex-bias
# and create a new variables assign X and autosomal linkage and sex
MALE_bias_df = merged[ ,c(4,7)]
head(MALE_bias_df)
FEMALE_bias_df = merged[ ,c(4,8)]
head(FEMALE_bias_df)

# MALE_bias_df
for(i in 1:nrow(MALE_bias_df)){  
  if(MALE_bias_df$chromosome[i] == "A"){   
    MALE_bias_df$chr_sex[i] <- "Auto_MALES"  
  } else {
    MALE_bias_df$chr_sex[i] <- "X_MALES"
  }
}

# sanity checks
str(MALE_bias_df$chr_sex)
MALE_bias_df$chr_sex <- as.factor(MALE_bias_df$chr_sex)
summary(MALE_bias_df$chr_sex)


# FEMALE_bias_df
for(i in 1:nrow(FEMALE_bias_df)){  
  if(FEMALE_bias_df$chromosome[i] == "A"){   
    FEMALE_bias_df$chr_sex[i] <- "Auto_FEMALES"  
  } else {
    FEMALE_bias_df$chr_sex[i] <- "X_FEMALES"
  }
}

# sanity checks
str(FEMALE_bias_df$chr_sex)
FEMALE_bias_df$chr_sex <- as.factor(FEMALE_bias_df$chr_sex)
summary(FEMALE_bias_df$chr_sex)


## 8. Rename the columns of the two subsets and then merge them
colnames(MALE_bias_df)<-c("chromosome","bias", "chr_sex") 
colnames(FEMALE_bias_df)<-c("chromosome","bias", "chr_sex") 
head(FEMALE_bias_df)

bias_df <- do.call("rbind", list(MALE_bias_df, FEMALE_bias_df))
# sanity checks
str(bias_df)
head(bias_df)

# 9. Calculate sex-bias percentages

# 9.1. female- and male-biased genes
Auto_fem_bias <- sum(bias_df$bias[which(bias_df$chr_sex=="Auto_FEMALES")])
Auto_fem_bias
X_fem_bias <- sum(bias_df$bias[which(bias_df$chr_sex=="X_FEMALES")])
X_fem_bias

Auto_male_bias <- sum(bias_df$bias[which(bias_df$chr_sex=="Auto_MALES")])
Auto_male_bias
X_male_bias <- sum(bias_df$bias[which(bias_df$chr_sex=="X_MALES")])
X_male_bias

# 9.2. female- and male-unbiased genes
# reminder of the total number of autosomal and X-lined genes for which GE values are available in this tissue
total_Auto_genes
total_X_genes

Auto_fem_unbias <- total_Auto_genes - Auto_fem_bias
Auto_fem_unbias
X_fem_unbias <- total_X_genes - X_fem_bias
X_fem_unbias

Auto_male_unbias <- total_Auto_genes - Auto_male_bias
Auto_male_unbias
X_male_unbias <- total_X_genes - X_male_bias
X_male_unbias

# 9.3. percentage of autosomal and X-linked female- and male-biased genes 
Auto_fem_bias_perc <- (Auto_fem_bias/total_Auto_genes)*100
X_fem_bias_perc <- (X_fem_bias/total_X_genes)*100
Auto_male_bias_perc <- (Auto_male_bias/total_Auto_genes)*100
X_male_bias_perc <- (X_male_bias/total_X_genes)*100

# 9.4. Gather these values in a dataframe do do the stats and plots
df <- data.frame(row.names = c("Auto_FEMALES","X_FEMALES","Auto_MALES","X_MALES"),
                 Chr_sex = c("Auto_FEMALES","X_FEMALES","Auto_MALES","X_MALES"), 
                 Chromosome = c("A","X","A","X"),
                 biased = c(Auto_fem_bias, X_fem_bias, Auto_male_bias, X_male_bias),
                 unbiased = c(Auto_fem_unbias, X_fem_unbias, Auto_male_unbias, X_male_unbias),
                 perc_bias = c(Auto_fem_bias_perc, X_fem_bias_perc, Auto_male_bias_perc, X_male_bias_perc))
                 
df

# 10. Stats - Chi-square test to compare the number of biased vs unbiased on the X and autosomes in each sex

# 10.1. subset only the 2 columns necessary for the test
df_for_chi_square = df [, c(3:4)]
df_for_chi_square

# 10.2. generate a matrix from it and perform a chi-square tests
Matriz = data.matrix(df_for_chi_square)
Matriz
stat.test <- pairwiseNominalIndependence(Matriz,
                                         fisher = FALSE,
                                         gtest  = FALSE,
                                         chisq  = TRUE,
                                         method = "fdr")
# all possible outcomes of the test
stat.test
# retain only the comparisons needed for the plot: Auto_Fem vs X_Fem and Auto_male vs X_male
stat.test.rows = stat.test [c(1,6), ]
head(stat.test.rows)
# tranform the dataframe to make it readable by the plot : split the comparison column into 2 new variables "groups"
stat.test.rows[c('group1', 'group2')] <- str_split_fixed(stat.test.rows$Comparison, ' : ', 2)
# delete the comparison column
stat.test.rows = stat.test.rows[, -1]
# re-organise the dataframe
stat.test.rows2 = stat.test.rows[c("group1", "group2", "p.Chisq" , "p.adj.Chisq")]
# make groups as factors
stat.test.rows2$group1 <- as.factor(stat.test.rows2$group1)
stat.test.rows2$group2 <- as.factor(stat.test.rows2$group2)
# out of the 2 p-value columns, keep adjusted
stat.test.rows2 = stat.test.rows2[, -3]
# rename columms
colnames(stat.test.rows2)<-c("group1", "group2", "p") 
head(stat.test.rows2)

# 11. Re-organise the levels of chr_sex variable for the plot
df$Chr_sex <- factor(df$Chr_sex, levels = c("Auto_FEMALES","X_FEMALES","Auto_MALES","X_MALES"))


# 12. Percentage sex-biased Plot
PLOT_bias <- ggplot(data = df, aes(y=perc_bias, x=Chr_sex, fill=Chromosome)) + 
  
  geom_bar(data = df, position="dodge", stat="identity") +
  
  scale_fill_manual(values = c("#969696","#ffd92f"), labels = c('Autosomes', 'X')) +
  
  coord_cartesian(ylim = c(0, 42))+
  ylab("Percentage of genes") +
  scale_y_continuous(breaks=seq(0,50,by=5), expand = c(0,0))+ #, expand = c(0, 0), limits = c(0, NA))+
  
  xlab("chromo_sex") + 
  
  ggtitle("Gonads - Percentage of sex-biased genes on X & autosomes \n (q < 0.05, TPM > 0.5, nfold = 2) \n (Test: chi-square)") +
  
  stat_pvalue_manual(data=stat.test.rows2, 
                     label.size=3.5,
                     inherit.aes=FALSE, 
                     y.position=c(38,16),
                     label = "p = {p}", 
                     fontface = "bold", 
                     bracket.size = 0.5, 
                     tip.length = 0)+
  
  theme(
    legend.position = c(.99, .99),
    legend.justification = c("right", "top"),
    legend.text=element_text(size=8),
    legend.title=element_blank(),
    
    axis.line = element_line(colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black",fill = NA,size = 1),
    panel.background = element_blank(),
    
    plot.title = element_text(hjust = 0, size=7, face = "bold"),
    
    axis.text.y = element_text(size=12),
    axis.title.y = element_text(margin=margin(r=3), size = 12, face ="bold"),
    
    axis.text.x = element_text(vjust = 1, hjust = 1, size = 10, face ="bold", color = "black", angle = 45),
    
    axis.title.x = element_blank())+
  
  scale_x_discrete(limits = c("Auto_FEMALES","X_FEMALES","Auto_MALES","X_MALES"),
                   breaks = c("Auto_FEMALES","X_FEMALES","Auto_MALES","X_MALES"),
                   labels=c('female_biased','female_biased', 'male_biased','male_biased'))


PLOT_bias


# 12. arrange plots in the same window
plots <- ggarrange(nfold_plot, PLOT_bias, nrow=1, ncol=2)
plots

# 13. save plots
#jpeg("GONADS_SBgenes_25scf_TPM05_nfold2_01062023.jpg", width = 7, height = 5, units = "in", res = 400)
#plots
#dev.off()


# 14. detach packages
detach("package:ggpubr", unload=TRUE)
detach("package:rcompanion", unload=TRUE)
detach("package:stringr", unload=TRUE)
detach("package:dplyr", unload=TRUE)
detach("package:EnvStats", unload=TRUE)
detach("package:ggplot2", unload=TRUE)
```
